src/twisted/web/_websocket_impl.py

```diff
@@ -0,0 +1,457 @@
+        """
+        Send a text message.
+        """
+
+    def sendBytesMessage(self, data: bytes) -> None:
+        """
+        Send a bytes message.
+        """
+
+    def loseConnection(self, code: int = 1000) -> None:
+        """
+        Drop the websocket connection.
+        """
+
+    def ping(self, payload: bytes = b"") -> None:
+        """
+        Send a websocket Ping request to measure latency.
+
+        @note: Per U{Mozilla's documentation
+            <https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#pings_and_pongs_the_heartbeat_of_websockets>},
+            multiple 'ping' requests may be coalesced into a single 'pong', and
+            unsolicited 'pong' requests must be ignored, so we do not return a
+            L{deferred <twisted.internet.defer.Deferred>} here; pongs are
+            delivered separately.
+        """
+
+    def attachProducer(self, producer: IPushProducer) -> None:
+        """
+        Attach the given L{IPushProducer} to this transport.
+        """
+
+    def detachProducer(self) -> None:
+        """
+        Detach a previously attached L{IPushProducer} from this transport.
+        """
+
+
+@dataclass
+class ConnectionRejected(Exception):
+    """
+    A websocket connection was rejected by an HTTP response.
+
+    @ivar response: The HTTP response that describes the rejection.
+    """
+
+    response: Response
+
+
+class WebSocketProtocol(TypingProtocol):
+    """
+    An object that conforms to L{WebSocketProtocol} can receive all the events
+    from a websocket connection.
+    """
+
+    def negotiationStarted(self, transport: WebSocketTransport) -> None:
+        """
+        An underlying transport (e.g.: a TCP connection) has been established;
+        negotiation of the websocket transport has begun.
+        """
+
+    def negotiationFinished(self) -> None:
+        """
+        Negotiation is complete: a bidirectional websocket channel is now fully
+        established.
+        """
+
+    def pongReceived(self, payload: bytes) -> None:
+        """
+        A 'pong' message was received.
+        """
+
+    def textMessageReceived(self, message: str) -> None:
+        """
+        A text message was received from the peer.
+        """
+
+    def bytesMessageReceived(self, data: bytes) -> None:
+        """
+        A bytes message was received from the peer.
+        """
+
+    def connectionLost(self, reason: Failure) -> None:
+        """
+        The websocket connection was lost.
+        """
+
+
+_WSP = TypeVar("_WSP", covariant=True, bound=WebSocketProtocol)
+_Bootstrap = Callable[[Union[WSConnection, Connection], ITransport], None]
+_log = Logger()
+
+
+class WebSocketServerFactory(TypingProtocol[_WSP]):
+    """
+    A L{WebSocketServerFactory} is a factory for a particular kind of
+    L{WebSocketProtocol} that implements server-side websocket listeners via
+    L{WebSocketResource}.
+    """
```

Comment:
How is this different from a normal factory ?

```suggestion
class WebSocketServerFactory(TypingProtocol[_WSP]):
    """
    A L{WebSocketServerFactory} is a factory for a particular kind of
    L{WebSocketProtocol} that implements server-side websocket listeners via
    L{WebSocketResource}.
    
    This implements the standard C{twisted.internet.protocol.Factory} API, but is provided here to help with type checking.
    """
```

Reply:
> This implements the standard C{twisted.internet.protocol.Factory} API, but is provided here to help with type checking.

It doesn't do that though; the signature is different, which is the whole reason that this class exists (and is in fact two classes)

Reply:
This is explicitly called out a bit more in the tutorial now.